go/extractor/cli/go-autobuilder/go-autobuilder.go

```diff
@@ -719,82 +719,150 @@ func installDependenciesAndBuild() {
 const minGoVersion = "1.11"
 const maxGoVersion = "1.20"
 
+// Check if `version` is lower than `minGoVersion`. Note that for this comparison we ignore the
+// patch part of the version, so 1.20.1 and 1.20 are considered equal.
+func belowSupportedRange(version string) bool {
+	return semver.Compare(semver.MajorMinor("v"+version), "v"+minGoVersion) < 0
+}
+
+// Check if `version` is higher than `maxGoVersion`. Note that for this comparison we ignore the
+// patch part of the version, so 1.20.1 and 1.20 are considered equal.
+func aboveSupportedRange(version string) bool {
+	return semver.Compare(semver.MajorMinor("v"+version), "v"+maxGoVersion) > 0
+}
+
 // Check if `version` is lower than `minGoVersion` or higher than `maxGoVersion`. Note that for
 // this comparison we ignore the patch part of the version, so 1.20.1 and 1.20 are considered
 // equal.
 func outsideSupportedRange(version string) bool {
-	short := semver.MajorMinor("v" + version)
-	return semver.Compare(short, "v"+minGoVersion) < 0 || semver.Compare(short, "v"+maxGoVersion) > 0
+	return belowSupportedRange(version) || aboveSupportedRange(version)
 }
 
-// Check if `v.goModVersion` or `v.goEnvVersion` are outside of the supported range. If so, emit
-// a diagnostic and return an empty version to indicate that we should not attempt to install a
-// different version of Go.
-func checkForUnsupportedVersions(v versionInfo) (msg, version string) {
-	if v.goModVersionFound && outsideSupportedRange(v.goModVersion) {
-		msg = "The version of Go found in the `go.mod` file (" + v.goModVersion +
-			") is outside of the supported range (" + minGoVersion + "-" + maxGoVersion +
-			"). Writing an environment file not specifying any version of Go."
-		version = ""
-		diagnostics.EmitUnsupportedVersionGoMod(msg)
-	} else if v.goEnvVersionFound && outsideSupportedRange(v.goEnvVersion) {
-		msg = "The version of Go installed in the environment (" + v.goEnvVersion +
-			") is outside of the supported range (" + minGoVersion + "-" + maxGoVersion +
-			"). Writing an environment file not specifying any version of Go."
+// Assuming `v.goModVersionFound` is false, emit a diagnostic and return the version to install,
+// or the empty string if we should not attempt to install a version of Go.
+func getVersionWhenGoModVersionNotFound(v versionInfo) (msg, version string) {
+	if !v.goEnvVersionFound {
+		// We definitely need to install a version. We have no indication which version was
+		// intended to be used to build this project. Go versions are generally backwards
+		// compatible, so we install the maximum supported version.
+		msg = "No version of Go installed and no `go.mod` file found. Writing an environment " +
+			"file specifying the maximum supported version of Go (" + maxGoVersion + ")."
+		version = maxGoVersion
+		diagnostics.EmitNoGoModAndNoGoEnv(msg)
```

Comment:
Are diagnostics put in the database? I thought they were just written as files somewhere, but I admit I don't know what happens after that.

Reply:
I now understand what you mean. If the diagnostic directory environment variable is not set, no diagnostics will be emitted, and one log line will be printed stating that. If we want to gather the diagnostics we'll have to set the env var and collect the diagnostic files ourselves in the action. This can be a future task.

Reply:
Makes sense -- if identify-environment can raise diagnostics, this could be included in its output for the Action to record once that makes sense.